typesetters/base.lua

```diff
@@ -330,6 +337,31 @@ function typesetter:endline ()
   SILE.documentState.documentClass.endPar(self)
 end
 
+-- Just compute once, to avoid unicode characters in source code.
+local speakerChangePattern = "^"
+   .. luautf8.char(0x2014) -- emdash
+   .. "[ " .. luautf8.char(0x00A0) .. luautf8.char(0x202F) -- regular space or NBSP or NNBSP
+   .. "]+"
+local speakerChangeReplacement = luautf8.char(0x2014) .. " "
+
+-- Special unshaped node subclass to handle space after a speaker change in dialogues
+-- introduced by an em-dash.
+local speakerChangeNode = pl.class(SILE.nodefactory.unshaped)
+function speakerChangeNode:shape()
+  local node = self._base.shape(self)
+  local spc = node[2]
+  if spc and spc.is_glue then
+    -- Switch the variable space glue to a fixed kern
+    node[2] = SILE.nodefactory.kern({ width = spc.width.length })
+    node[2].parent = self.parent
+  else
+    -- Should not occur:
+    -- How could it possibly be shaped differently?
+    SU.warn("Speaker change logic met an unexpected case, this might be a bug.")
```

Comment:
If this is "Should not occur" error then `SU.error("...", true)` would be more appropriate to actually throw a bug error.

Reply:
My rationale for a warning is that it won't really break (but it will just leave a stretchable glue): so we should perhaps not block users here if the edge-case is ever met for any reason -- it's not worse than what they'd be getting with the feature disabled. It can be changed to an error if deemed preferable (and if the case occurs, we'll need to be on the front-line for fixing it)